cmd/thanos/endpointset.go

```diff
@@ -0,0 +1,330 @@
+	}
+	res.cfg = cfg
+
+	reloadCtx, reloadCancel := context.WithCancel(context.Background())
+	g.Add(func() error {
+		return extkingpin.PathContentReloader(reloadCtx, configFile, logger, func() {
+			res.mu.Lock()
+			defer res.mu.Unlock()
+
+			level.Info(logger).Log("msg", "reloading endpoint config")
+			cfg, err := res.parse(configFile)
+			if err != nil {
+				level.Error(logger).Log("msg", "failed to reload endpoint config", "err", err)
+				return
+			}
+			res.cfg = cfg
+		}, configReloadInterval)
+	}, func(err error) {
+		reloadCancel()
+	})
+
+	return res, nil
+}
+
+type configProvider interface {
+	config() EndpointConfig
+}
+
+type providerFunc func() EndpointConfig
+
+func (f providerFunc) config() EndpointConfig { return f() }
+
+// For querier
+func setupEndpointset(
+	g *run.Group,
+	reg prometheus.Registerer,
+	logger log.Logger,
+	configFile *extflag.PathOrContent,
+	configReloadInterval time.Duration,
+	dnsSDResolver string,
+	dnsSDInterval time.Duration,
+	unhealthyTimeout time.Duration,
+	endpointTimeout time.Duration,
+	queryConnMetricLabels ...string,
+) (*query.EndpointSet, error) {
+	reloader, err := newEndpointReloader(g, logger, configFile, configReloadInterval)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to load config initially")
+	}
+	return setupEndpointsetWithProvider(
+		g,
+		reg,
+		logger,
+		reloader,
+		dnsSDResolver,
+		dnsSDInterval,
+		unhealthyTimeout,
+		endpointTimeout,
+		queryConnMetricLabels...,
+	)
+}
+
+// For ruler ( for some reason )
+func setupStaticEndpointset(
+	g *run.Group,
+	reg prometheus.Registerer,
+	logger log.Logger,
+	endpointCfg EndpointConfig,
+	dnsSDResolver string,
+	dnsSDInterval time.Duration,
+	unhealthyTimeout time.Duration,
+	endpointTimeout time.Duration,
+	queryConnMetricLabels ...string,
+) (*query.EndpointSet, error) {
+	provider := providerFunc(func() EndpointConfig {
+		return endpointCfg
+	})
+	return setupEndpointsetWithProvider(
+		g,
+		reg,
+		logger,
+		provider,
+		dnsSDResolver,
+		dnsSDInterval,
+		unhealthyTimeout,
+		endpointTimeout,
+		queryConnMetricLabels...,
+	)
+}
+
+func setupEndpointsetWithProvider(
+	g *run.Group,
+	reg prometheus.Registerer,
+	logger log.Logger,
+	cfgProvider configProvider,
+	dnsSDResolver string,
+	dnsSDInterval time.Duration,
+	unhealthyTimeout time.Duration,
+	endpointTimeout time.Duration,
+	queryConnMetricLabels ...string,
+) (*query.EndpointSet, error) {
+	// Register resolver for the "thanos:///" scheme for endpoint-groups
+	dns.RegisterGRPCResolver(
+		dns.NewProvider(
+			logger,
+			extprom.WrapRegistererWithPrefix("thanos_query_endpoint_groups_", reg),
+			dns.ResolverType(dnsSDResolver),
+		),
+		dnsSDInterval,
+	)
+	dnsEndpointProvider := dns.NewProvider(
+		logger,
+		extprom.WrapRegistererWithPrefix("thanos_query_endpoints_", reg),
+		dns.ResolverType(dnsSDResolver),
+	)
+	duplicatedStores := promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_query_duplicated_store_addresses_total",
```

Comment:
This code path is shared by ruler/query so this changes the metric name. I think we will need to pass `query`/`rule` string to this function and use it in the metric's name

Reply:
Ill add a component.Component!

Reply:
added component